docs/rfcs/rfc03.md

```diff
@@ -0,0 +1,127 @@
+# RFC02: Caching and vendoring
+
+## Background
+
+Jsonnet goes way out of its way to ensure that all the imports are well known statically, and that the evaluation is fully reproducible
+and cacheable by providing side-effect free evaluation and avoiding computed imports.
+
+Kubecfg adds the ability to directly import external dependencies. This is meant to be used for importing immutable external files,
+usually links to some resource at some specific git sha. It also [supports](https://github.com/kubecfg/kubecfg/pull/218) explicitly enforcing a fingerprint check.
+
+However, fetching external dependencies comes with problems:
+
+* downloding external dependencies may be slow and annoying
+* external files may disappear permanently or temporarily
+
+## Proposal
+
+We propose adding a download cache that will store a local copy of files downloaded from remote sources.
+
+The cache directory structure, is suitable for checking-in the cache into the current VCS repo (if the user so chooses).
+
+### Cache structure
+
+Let's consider this jsonnet source file:
+
+```jsonnet
+import 'https://github.com/kubecfg/k8s-libsonnet/raw/144e62ccbdc9f7d4f8bccc14f33224cbf63e5185/k8s.libsonnet'
+```
+
+The cache directory can be set via the `--cache-dir` flag. By default its `<cachebase>/kubecfg`, where `<cachebase>` is determined by the [os.UserCacheDir](https://pkg.go.dev/os@go1.20.2#UserCacheDir).
+
+
+After the first execution, kubecfg will create this directory structure
+
+```console
+$ tree ~/Library/Caches/kubecfg
+/Users/mkm/Library/Caches/kubecfg
+└── https
+    └── github.com
+        └── kubecfg
+            └── k8s-libsonnet
+                └── raw
+                    └── 144e62ccbdc9f7d4f8bccc14f33224cbf63e5185
+                        └── k8s.libsonnet
+```
+
+### Multiple cache sources
+
+(optional, advanced)
+
+The user can define multiple cache directories.
+They will act as a "search path".
+If a file is not found in the first directory, the next one will be tried until the cache directories are exhausted.
+Only the first cache directory is used to write new files to.
+
+Example:
+
+```bash
+kubecfg --cache-dir ./vendor --cache-dir ~/.cache/kubecfg show ....
+```
+
+Open questions:
+
+* should we always put the default cache at the end of the cache search path?
+
+### Vendoring
+
+To enable vendoring the user needs to pass a custom location for the cache directory and choose a desirable cache pruning policy.
```

Comment:
The thing that I would really like is a package manager. But as I started writing that comment, it occurred to me that consumers could probably build their own out of these primitives. And the more jsonnet-ey solution might be to let them do that, with some guidance as to how.

The way I'm imaging that would work is to create something like a packages.jsonnet file, and that would import and expose all of the external dependencies. Then, as a matter of convention, the rest of the jsonnet project references the objects exposed by that file, and never directly imports their own external dependencies.

Like so
```jsonnet
//<project_root>/packages.jsonnet
{
    acmeAppSpec: import 'https://github.com/acme/exampleApp/<gitsha>'
}
```

```jsonnet
//<project_root>/env/production.jsonnet
{
    app: k8s.apps.v1.Deployment {
        spec+: (import '../packages.jsonnet').acmeAppSpec {
            // normal jsonnet things here
        }
    }
}
```

Then if something like this becomes the standard guidance, the rest of the tooling can be designed to support that. For instance, cache cleaning can remove unused references, and it's easy to know what's used or not based just on this one file. If a repo needs to have more than one set of external references for some reason, each set can be expressed with it's own packages files, and its own cache dir.

Reply:
I think this also supports more advanced usage, too. For example
```jsonnet
//<project_root>/packages.jsonnet
{
    acmeAppSpec: {
        v2Config:: { experimental: true }
        v1: import 'https://github.com/acme/exampleApp/<gitsha>/app.jsonnet',
        v2: (import 'https://github.com/acme/appTwoDotOh/<gitsha>/app.jsonnet') + self.v2Config,
    }
}
```

Reply:
Not quite sure if we need a "package manager". If we have a way to point to remote sources a tool like renovate bot can just update those references. If you don't want to have them scattered across your files, you can easily create a level of indirection:

```
$ cat packages.jsonnet
{
  foo: import 'https://github.com/acme/foo/v0.1.0/app.jsonnet',
  bar: import 'https://github.com/acme/bar/v0.3.0/app.jsonnet',  
}
$ cat something.jsonnet
{
  foo: (import 'packages.jsonnet').foo,
}
```

then managing the versions of your dependencies across your project is just a matter of updating `packages.jsonnet` and you have all foos point to the same foo, while at the same time allowing people to import two different versions of a dependency if they really need it (something is quite a pain nowadays with jsonnet bundler)